### PR TITLE
[refs] `lib.convert` preserves `:ident`s through a round trip

### LINKED ISSUE #49682: Modify legacy MBQL to preserve the randomly-generated `:ident`s of aggregations, join clauses, etc.

### LINKED ISSUE #49683: Update `lib.convert` to preserve `:ident`s

### PR DESCRIPTION
**Pull request by bshepherdson**
This is the second edition of #50085, reverted by #51128. Closes #49682. Closes #49683.

Updates `lib.convert` and legacy MBQL normalization to allow
`:ident`s, and fixes some QP middleware and utils so they can
still match up refs correctly.

No user-visible impact.

Newly created queries will have `:ident`s generated for joins,
aggregations, etc. by the previous PR. This starts saving them
into appDBs. It is still acceptable to have missing idents at
this stage, including for new queries created without the lib.

Pre-existing clauses don't have refs added yet.

### Differences from the original PR

- Relaxed the schemas to make `:ident`s optional (on joins) and nullable (in `:aggregation-idents` et al).
- Now things work fine without idents, so the e2e tests are unchanged (and stats won't break)

### Contrib

* Fix loading in segments
    - Co-authored-by: Alexander Polyankin <alexander.polyankin@metabase.com>

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by bshepherdson** [comment]
* **#51133** <a href="https://app.graphite.dev/github/pr/metabase/metabase/51133?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/51133?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: dev/src/dev/portal.clj [hunk 1]
```diff
@@ -79,11 +79,26 @@
...
     (if forms
       (let [form     (first forms)
-            threaded `(let [before# ~x
-                            after#  (-> before# ~form)]
-                        (tap> [~(list `quote form)
-                               ^{:portal.viewer/default :portal.viewer/diff}
-                               [before# after#]])
-                        after#)]
+            threaded (let [args       (when (and (list? form)
+                                                 (seq (rest form)))
```
Comment (metamben) [inline dev/src/dev/portal.clj:83]:
```suggestion
                                                 (next form))
```

File: src/metabase/legacy_mbql/normalize.cljc [hunk 2]
```diff
@@ -399,12 +400,21 @@
...
-                     :expressions     normalize-expressions-tokens
-                     :order-by        normalize-order-by-tokens
-                     :source-query    normalize-source-query
-                     :source-metadata {::sequence normalize-source-metadata}
-                     :joins           {::sequence normalize-join}}
+   :query           {:aggregation        normalize-ag-clause-tokens
+                     :aggregation-idents (fn [m]
+                                           (update-keys m #(cond-> %
+                                                             (string? %) parse-long
+                                                             (keyword? %) (-> name parse-long))))
```
Comment (metamben) [inline src/metabase/legacy_mbql/normalize.cljc:407]: Would this be too tricky?
```suggestion
                                                             (keyword? %) name
                                                             (string? %) parse-long)))
```

**Review by lbrdnk — APPROVED** [review]

File: src/metabase/xrays/automagic_dashboards/combination.clj [hunk 3]
```diff
@@ -38,10 +38,19 @@
   [query
    breakout-fields
    filter-clauses]
-  (cond->
-   (assoc query :breakout (mapv (partial interesting/->reference :mbql) breakout-fields))
-    (seq filter-clauses)
-    (assoc :filter (into [:and] filter-clauses))))
+  (let [breakouts  (mapv (partial interesting/->reference :mbql) breakout-fields)
+        brk-idents (into {} (for [i (range (count breakouts))]
+                              [i (u/generate-nano-id)]))]
```
Comment (metamben) [inline src/metabase/xrays/automagic_dashboards/combination.clj:43]: I find the `zipmap` pattern you used before easier to read.
```suggestion
        brk-idents (zipmap (range (count breakouts)) (repeatedly u/generate-nano-id))]
```

File: src/metabase/xrays/transforms/core.clj [hunk 4]
```diff
@@ -69,26 +69,34 @@
...
+          (assoc :expressions       expr-clauses
+                 :expression-idents (update-vals expr-clauses (fn [_] (u/generate-nano-id))))
+          (update :fields concat (for [expression (keys expressions)]
+                                   [:expression expression]))))
     query))
 
+(defn- indexed-idents [seqable]
+  (when (seq seqable)
+    (into {} (map (fn [i] [i (u/generate-nano-id)]))
+          (range (count seqable)))))
```
Comment (metamben) [inline src/metabase/xrays/transforms/core.clj:85]: This is at least the third way of writing this. :)

**Review by metamben — APPROVED** [review]
LGTM

**Review by ranquild — APPROVED** [review]

File: src/metabase/xrays/transforms/core.clj (hunk 4 above)
Comment (bshepherdson) [reply src/metabase/xrays/transforms/core.clj:85]: I agree... it feels like this should be a built-in, sort of like `frequencies`.

File: src/metabase/legacy_mbql/normalize.cljc (hunk 2 above)
Comment (bshepherdson) [reply src/metabase/legacy_mbql/normalize.cljc:407]: A little... I'd rather keep them as separate clauses.

Comment (metamben) [reply src/metabase/legacy_mbql/normalize.cljc:407]: Then maybe some kind of `cond` would be more explicit.

File: src/metabase/xrays/transforms/core.clj (hunk 4 above)
Comment (bshepherdson) [reply src/metabase/xrays/transforms/core.clj:85]: I've consolidated these with a helper.

**Pull request merged**